### PR TITLE
Improves guidelines

### PR DESCRIPTION
**Pull request by andresgalante**
Hi @matthewcarleton, these are the changes on the guidelines we've talked about. I've changed of how we display in the information, I hope it makes more sense now.

Can you review it when you have time please?

### CODE REVIEW

File: CODE-GUIDELINES.md [hunk 1]
```diff
@@ -4,285 +4,68 @@ Please enforce these guidelines at all times. Small or large, call out what's in
 
 > Every line of code should appear to be written by a single person, no matter the number of contributors.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: Do we want to cite this quote?

File: CODE-GUIDELINES.md [hunk 2]
```diff
@@ -4,285 +4,68 @@ Please enforce these guidelines at all times. Small or large, call out what's in
...
+  - [Exceptions](#exceptions)
+  - [Deviations](#deviations)
+    - [HTML](#html)
+    - [Lintable HTML rules](#lintable-html-rules)
+    - [Comment and Organization](#comment-and-organization)
+  - [Additions](#additions)
+    - [Lintable CSS rules](#lintable-css-rules)
+    - [Shorthand notation](#shorthand-notation)
+    - [Typography](#typography)
+    - [Spaces, margins and paddings](#spaces--margins-and-paddings)
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: is it paddings or padding? I've never heard paddings before.

File: CODE-GUIDELINES.md [hunk 3]
```diff
@@ -4,285 +4,68 @@ Please enforce these guidelines at all times. Small or large, call out what's in
...
-  height: 15px;
-  background-image: url(../img/sprite.png);
-}
+**Practicality over purity**. Strive to maintain HTML standards and semantics, but not at the expense of practicality. Use the least amount of markup with the fewest intricacies whenever possible.
 
-.icon           { background-position: 0 0; }
-.icon--home      { background-position: 0 -20px; }
-.icon--account   { background-position: 0 -40px; }
-```
+The CSS linter is PatternFly's single source of truth for CSS syntax, declaration order and other CSS rules like disallow type selectors, disallow vendor prefixes, and more.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: Should we write the rules in bullet point to make them more obvious?

File: CODE-GUIDELINES.md [hunk 4]
```diff
@@ -4,285 +4,68 @@ Please enforce these guidelines at all times. Small or large, call out what's in
...
-  - [Prefixed properties](#prefixed-properties)
-  - [Single declarations](#single-declarations)
-  - [Shorthand notation](#shorthand-notation)
-  - [Comment and Organization](#comment-and-organization)
-  - [Naming Selectors](#naming-selectors)
-  - [Specificity](#specificity)
-  - [Spacing](#spacing)
-  - [Shadows](#shadows)
-  - [Gradients](#gradients)
+This set of rules generate some constraints and conventions. If you ran into instances where a convention isn’t obvious or a solution could be handled in a few different ways, contact the PatternFly community, have a conversation about how to handle it and update this guidelines when needed.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: run instead of ran

File: CODE-GUIDELINES.md [hunk 5]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
-### 1. License header
+##### 1. License header
 
 PatternFly is license under the [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0).
 
 - Always add the Apache v2.0 license header at the top of each scss files.
 - Leave three blank lines bellow.
 
-### 2. DocBlock
+##### 2. Doc Block
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: there is a double return below Doc Block

File: CODE-GUIDELINES.md [hunk 6]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
+  margin-bottom: 10px;
+  background-color: red;
+  background-image: url("image.jpg");
+  border-top-left-radius: 3px;
+  border-top-right-radius: 3px;
 }
 ```
 
-### IDs and classes
+The Mozilla Developer Network has a great article on [shorthand properties](https://developer.mozilla.org/en-US/docs/Web/CSS/Shorthand_properties) for those unfamiliar with notation and behaviour. And also [Harry Robert's](https://csswizardry.com/2016/12/css-shorthand-syntax-considered-an-anti-pattern/) article.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: This sounds a bit awkward, How about "The Mozilla Developer Network and Harry Robert's both have great articles on shorthand properties for those unfamiliar with notation and behaviour" Link the names.

File: CODE-GUIDELINES.md [hunk 7]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
-  h2{
-    font-size: 2em;
-  }
-}
-
-// Good
-.pf-card { color: #000; }
-
-.pf-card--title{ font-size: 2em; }
+Every typography element has a zero margin and base `font-size`. Margins and font treatments are added width classes.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: with instead of width

File: CODE-GUIDELINES.md [hunk 8]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
-- Each class name imparts structural info without binding to exact HTML.
-
-
-
-<!-- ============================================================ -->
-
-## Specificity
-
-Always keep selectors as shallow as possible.
+Always them to define `margin` and `padding`
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: Always use them

File: CODE-GUIDELINES.md [hunk 9]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
-1. States, pseudo-classes and pseudo-elements
-1. Overwrite Bootstrap
+```html
+<p class="mb-lg">This paragraph has a large bottom margin</p>
+```
 
+##### Comment all magic numbers
 
-For longer style blocks don't nest the modifier code as it reduced the legibility of the code.
+If a situation arises where a value needs entering into the authoring style sheets that isn't already defined in the variables this should serve as a red flag to you.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: Do you think we need to better define magic number here? We don't explicitly define it.

File: CODE-GUIDELINES.md [hunk 10]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
+<p class="pf-box-shadow-md">This paragraph has a medium box shadow</p>
 
-States of a component should be included as a nested element. This includes hover, focus, and active states:
+<button class="btn btn-primary pf-box-shadow-none">This button has no shadow</button>
+```
 
-```sass
-.btn {
-    background: blue;
+#### Gradients
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: should we spell out buttons here?

File: CODE-GUIDELINES.md [hunk 11]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
 ```
 
-#### 5. Overwrite Bootstrap
+### Sass
 
-Keep PatternFly code DRY. Reuse as much as you can from Bootstrap.
+PatternFly uses [Sass](http://sass-lang.com/) to preprocess CSS.
 
-**To overwrite Bootstrap:**
+As a general rule don't overcomplicate Sass, keep it easy to parse for normal human.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: "for a normal human", or "normal humans"

File: CODE-GUIDELINES.md [hunk 12]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
-- Be precise and accurate, introduce as little modifications as possible to achieve the design.
-- Base your styles on the original bootstrap scss file.
-- Bootstrap uses a modified version of BEM: `{{block}}-{{element}}-{{modifier}}`.
+#### Nesting
 
-```sass
-// PatternFly card block
-.pf-card {
-  border: none;
+Avoid Sass nesting, but if your really have to do it never go more than three layers deep.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: Avoid Sass nesting, if absolutely necessary don't exceed a three layer depth.

File: CODE-GUIDELINES.md [hunk 13]
```diff
@@ -327,386 +110,290 @@ Follow this comment structure:
...
 
-```html
-<p class="mb-lg">This paragraph has a large bottom margin</p>
-```
-
-<!-- ============================================================ -->
+Component-specific media queries should be nested inside the component block. Use [Bootstrap responsive breakpoints mixins](http://v4-alpha.getbootstrap.com/layout/overview/#responsive-breakpoints) and remember that PatternFly is mobile first.
 
-## Shadows
+**Do progressive enhancement, not gracefully degrade.**
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: degradation

File: CODE-GUIDELINES.md [hunk 14]
```diff
@@ -773,7 +459,105 @@ Treat @extend with respect. Use @extend only for maintaining relationships withi
...
 
-# Credits and references
+
+## How to write PatternFly selectors
+
+Since PatternFly is based on Bootstrap, we follow some rules to keep our library isolated, easy to maintain and grow.
+
+### Namespace
+
+To avoid conflicts PatternFly prefixes all classes with “pf-”. This also helps differentiates PatternFly and Bootstrap class names.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: ifferentiate

File: CODE-GUIDELINES.md [hunk 15]
```diff
@@ -773,7 +459,105 @@ Treat @extend with respect. Use @extend only for maintaining relationships withi
...
+UI structures fall into 4 categories:
+
+- Layouts
+- Components
+- Utilities
+- Bootstrap components
+
+#### Layout
+
+Layout are containers concern about it's children vertical and horizontal alignment and spacing.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: Do you want this line still, it seems to be repeated on the next line.

File: CODE-GUIDELINES.md [hunk 16]
```diff
@@ -773,7 +459,105 @@ Treat @extend with respect. Use @extend only for maintaining relationships withi
...
+
+Layout are containers concerned with the vertical and horizontal alignment and spacing of their children. The classes are prefixed with `-l` (after the patterfly prefix `pf-`), for example: `.pf-l-grid` or `.pf-layout-list`.
+
+#### Components
+
+Components are modular and independent structures concerned about how a thing looks.
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: concerned with

File: CODE-GUIDELINES.md [hunk 17]
```diff
@@ -773,7 +459,105 @@ Treat @extend with respect. Use @extend only for maintaining relationships withi
...
+Utility classes are prefixed with `-u` for example `.pf-u-shadow-sm`.
+
+
+#### Bootstrap components
+
+Keep PatternFly styles independent from Bootstrap.
+
+When writing a PatternFly component **don't extend or build on top of Bootstrap structure**.
+
+**Follow this rules to style Bootstrap components:**
```
Comment (matthewcarleton) [inline CODE-GUIDELINES.md]: these

File: CODE-GUIDELINES.md (hunk 3 above)
Comment (andresgalante) [reply CODE-GUIDELINES.md]: I'd say no. these are just examples not a list of the exact things we lint

File: CODE-GUIDELINES.md (hunk 10 above)
Comment (andresgalante) [reply CODE-GUIDELINES.md]: I'll leave this open till we have the new designs

**Pull request merged**